sky/data/sky_csync.py

```diff
@@ -0,0 +1,545 @@
+"""CSYNC module"""
+import functools
+import os
+import shlex
+import shutil
+import subprocess
+import sys
+import tempfile
+import time
+from typing import Any, List, Optional, Tuple
+
+import click
+import psutil
+
+from sky import exceptions
+from sky import sky_logging
+from sky.data import mounting_utils
+from sky.data.storage_utils import StorageMode
+from sky.skylet import constants
+from sky.utils import common_utils
+from sky.utils import db_utils
+
+logger = sky_logging.init_logger(__name__)
+
+_CSYNC_DB_PATH = os.path.join(constants.CSYNC_DIR, 'sky_csync.db')
+_CSYNC_READ_PATH = os.path.join(constants.CSYNC_DIR, 'read_{pid}')
+_CSYNC_WRITE_PATH = os.path.join(constants.CSYNC_DIR, 'write_{pid}')
+
+_BOOT_TIME = None
+_CURSOR = None
+_CONN = None
+_DB = None
+_MAX_SYNC_RETRIES = 10
+# Allowing minimal seconds of interval between SYNC calls.
+_MIN_SYNC_CALL_INTERVAL = 3
+_DEFAULT_S3_NUM_THREAD = 10
+CSYNC_DEFAULT_INTERVAL_SECONDS = 600
+
+
+def connect_db(func):
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        global _DB, _CURSOR, _CONN, _BOOT_TIME
+
+        def create_table(cursor, conn):
+            cursor.execute("""\
+                CREATE TABLE IF NOT EXISTS running_csync (
+                csync_pid INTEGER PRIMARY KEY,
+                sync_pid INTEGER DEFAULT -1,
+                storage_mount_pid INTEGER,
+                redirect_mount_pid INTEGER,
+                mountpoint_path TEXT,
+                boot_time FLOAT)""")
+
+            conn.commit()
+
+        if _DB is None:
+            db_path = os.path.expanduser(_CSYNC_DB_PATH)
+            os.makedirs(os.path.abspath(os.path.expanduser(
+                constants.CSYNC_DIR)),
+                        exist_ok=True)
+            _DB = db_utils.SQLiteConn(db_path, create_table)
+
+        _CURSOR = _DB.cursor
+        _CONN = _DB.conn
+        _BOOT_TIME = psutil.boot_time()
```

Comment:
add a indent for them to inside the `if _DB is None` block?

Reply:
What if there's multiple `CSYNC` mode storage used when running `sky launch`? Perhaps, should I leave out the `_BOOT_TIME`?

Reply:
Multiple csync mode will call the cli multiple times, so iiuc they won't affect each other. Here I mean the code will assign the `_CURSOR` and other variables everytime when the decorator is used (inside a single csync CLI call), which is not ideal